src/libraries/System.Private.CoreLib/src/System/Numerics/TotalOrderIeee754Comparer.cs

```diff
@@ -0,0 +1,141 @@
+        /// <param name="x">The first number to compare.</param>
+        /// <param name="y">The second number to compare.</param>
+        /// <returns>
+        /// A signed integer that indicates the relative
+        /// values of <paramref name="x"/> and <paramref name="y"/>.
+        /// </returns>
+        /// <remarks>
+        /// IEEE754 specification defines totalOrder as a &lt;= semantic.
+        /// totalOrder(x,y) is true corresponds to the result of this method &lt;= 0.
+        /// </remarks>
+        public int Compare(T? x, T? y)
+        {
+            // IComparer contract is null < value
+
+            if (x is null && y is null)
+            {
+                return 0;
+            }
+            else if (x is null)
+            {
+                return -1;
+            }
+            else if (y is null)
+            {
+                return 1;
+            }
+
+            // If < or > returns true, the result satisfies definition of totalOrder too
+
+            if (x < y)
+            {
+                return -1;
+            }
+            else if (x > y)
+            {
+                return 1;
+            }
+            else if (x == y)
+            {
+                if (T.IsZero(x) && T.IsZero(y))
+                {
+                    if (T.IsNegative(x) && T.IsPositive(y))
+                    {
+                        return -1;
+                    }
+                    else if (T.IsPositive(x) && T.IsNegative(y))
+                    {
+                        return 1;
+                    }
+                    else
+                    {
+                        return 0;
+                    }
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+            else
+            {
+                // One or two of the values are NaN
+                // totalOrder defines that -qNaN < -sNaN < x < +sNaN < + qNaN
+
+                if (T.IsNaN(x) && T.IsNaN(y))
+                {
+                    if (T.IsNegative(x) && T.IsPositive(y))
+                    {
+                        return -1;
+                    }
+                    else if (T.IsPositive(x) && T.IsNegative(y))
+                    {
+                        return 1;
+                    }
+                    else
+                    {
+                        // The order of same category of NaN is undefined
+                        return 0;
+                    }
+                }
+                else if (T.IsNaN(x))
+                {
+                    return T.IsPositive(x) ? 1 : -1;
+                }
+                else if (T.IsNaN(y))
+                {
+                    return T.IsPositive(y) ? -1 : 1;
+                }
+                else
+                {
+                    // T does not correctly implement IEEE754 semantics
+                    // return 0 for this case
+                    return 0;
+                }
```

Comment:
We want to `throw` if the input claims to be IEEE 754 compliant but ends up with some invalid comparison.

We should probably just use `ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidArgumentForComparison);` since that's what `Comparer<T>` uses for an invalid `T`

Reply:
If I'm understanding the binary fp encoding correctly, isn't totalOrder comparing all the bits except sign in integer semantic?
For normal values, larger exponent always corresponds to larger values since there is an implicit `1.` before the significant bits.
Inf and NaN are all setting exponent to 1s, Inf sets significant to 0, sNaN sets to 0xxx, and qNaN sets to 1xxx, and totalOrder defines Inf<sNaN<qNaN.

It's really attractive for me to write the comparison in integers, however there looks no way to do this in generic math, and I'm not sure about decimals.

Reply:
You'd need to add specialized paths such as:
```csharp
if (typeof(T) == typeof(float))
{
    return CompareSingle((float)(object)(left), (float)(object)(right));
}
else if (typeof(T) == typeof(double))
{
    return CompareDouble((double)(object)(left), (double)(object)(right));
}
else if (typeof(T) == typeof(Half))
{
    return CompareHalf((Half)(object)(left), (Half)(object)(right));
}
else
{
    return CompareGeneric(left, right);
}
```

Reply:
Then there would be no direct test coverage for generic version. What's the best approach here? Making synthesized delegating types in test code?

Reply:
You can test it using `System.Runtime.CompilerServices.NFloat` -or- create a fake type that implements `IFloatingPointIeee754` and minimally wraps `float`, etc.